### PR TITLE
[Doc] Update Customize Datasets Doc

### PR DESCRIPTION
**Pull request by MengzhangLI**

### CODE REVIEW

File: docs/en/advanced_guides/add_dataset.md [hunk 1]
```diff
@@ -1,5 +1,54 @@
...
+        type=dataset_type,
+        data_root=data_root,
+        data_prefix=dict(
+            img_path='images/validation',
+            seg_map_path='annotations/validation'),
+        pipeline=test_pipeline))
+test_dataloader = val_dataloader
+```
+
+- `train_dataloader`, `val_dataloader` and `test_dataloader`: The [`config`](https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/config.md)s to build dataset instances for model training, validation and testing by
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:36]: config is moved in engine from mmcv

Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:36]: The setting of dataloaders  works for model training, validation and testing that are by `build and registry` mechanism

File: docs/en/advanced_guides/add_dataset.md [hunk 2]
```diff
@@ -1,5 +1,54 @@
...
+  using [`build and registry`](https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/registry.md) mechanism.
+
+- `batch_size`: How many samples per batch and per gpu to load during model training, and the total batch sizes of training is equal to `batch_size` times gpu number, e.g. when using 8 gpus for distributed data parallel trainig and `batch_size=4`, the total batch sizes is `8*4=32`.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:39]: How many samples per batch and per gpu to load during model training - > the number of a batch sample of each gpu

File: docs/en/advanced_guides/add_dataset.md [hunk 3]
```diff
@@ -1,5 +1,54 @@
...
+  `val_dataloader` with mmseg >=0.24.1.
+
+- `num_workers`: How many subprocesses per gpu to use for data loading. `0` means that the data will be loaded in the main process.
+
+**Note:** Default setting of `batch_size` is 1 in MMSegmentation when model testing and validation (DO NOT support batch inference yet).
+
+**Note:** Before v0.24.1, except `train`, `val` `test`, `samples_per_gpu` and `workers_per_gpu`, the other keys in `data` must be the
+input keyword arguments for `dataloader` in pytorch, and the dataloaders used for model training, validation and testing have the same input arguments.
+In v0.24.1, MMSegmentation supports to use `train_dataloader`, `test_dataloaser` and `val_dataloader` to specify different keyword arguments, and still supports the overall arguments definition but the specific dataloader setting has a higher priority.
+After v1.0.0rc0, MMSegmentation would only use `train_dataloader`, `test_dataloaser` and `val_dataloader`.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:50]:
```suggestion
```

File: docs/en/advanced_guides/add_dataset.md [hunk 4]
```diff
@@ -1,5 +1,54 @@
...
+- `train_dataloader`, `val_dataloader` and `test_dataloader`: The [`config`](https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/config.md)s to build dataset instances for model training, validation and testing by
+  using [`build and registry`](https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/registry.md) mechanism.
+
+- `batch_size`: How many samples per batch and per gpu to load during model training, and the total batch sizes of training is equal to `batch_size` times gpu number, e.g. when using 8 gpus for distributed data parallel trainig and `batch_size=4`, the total batch sizes is `8*4=32`.
+  If you would like to define `batch_size` for testing and validation, please use `test_dataloaser` and
+  `val_dataloader` with mmseg >=0.24.1.
+
+- `num_workers`: How many subprocesses per gpu to use for data loading. `0` means that the data will be loaded in the main process.
+
+**Note:** Default setting of `batch_size` is 1 in MMSegmentation when model testing and validation (DO NOT support batch inference yet).
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:45]:
```suggestion
**Note:** Default setting of `batch_size` is `1` in MMSegmentation when model testing and validation due to different shapes of samples in a dataset.
```

File: docs/en/advanced_guides/add_dataset.md [hunk 5]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
 For example, suppose the original dataset is `Dataset_A`, to repeat it, the config looks like the following
 
 ```python
-dataset_A_train = dict(
+dataset_A_train_dataloader = dict(
+    dataset=dict(
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:111]: there is a configuration for dataset.

File: docs/en/advanced_guides/add_dataset.md [hunk 6]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+_base_ = [
+    '../_base_/models/pspnet_r50-d8.py', '../_base_/datasets/my_dataset_config.py',
+    '../_base_/default_runtime.py', '../_base_/schedules/schedule_80k.py'
+]
+model = dict(
+    decode_head=dict(num_classes=YOUR_DATASET_CLASSES), auxiliary_head=dict(num_classes=YOUR_DATASET_CLASSES))
+
+```
+
+By now, you can use your own dataset in MMSegmentation.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:372]:
```suggestion
```

File: docs/en/advanced_guides/add_dataset.md [hunk 7]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+
+After your dataset conversion done above, create a new file `my_dataset.py` in `./mmseg/dataset/`, where defining `MyDataset` class could be registered into `DATASETS` of MMCV and be used by model.
+
+```python
+from .builder import DATASETS
+from .custom import CustomDataset
+
+# Register MyDataset class into DATASETS
+@DATASETS.register_module()
+class MyDataset(CustomDataset):
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:262]: there is no customdataset in mmseg

File: docs/en/advanced_guides/add_dataset.md [hunk 8]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+            img_suffix='.png',
+            seg_map_suffix='.png',
+            reduce_zero_label=False, # reduce_zero_label is False because label 0 is background (first one in CLASSES above)
+            **kwargs)
+```
+
+`classes` and `palette` in `METAINFO` are defined in `./mmseg/dataset/my_dataset.py`, which are class names and BGR values of annotations, respectively.
+Besides,`palette` would only be used in prediction visulization, which would not affect process of training and validation.
+
+Specifically, if label 0 in segmentation map is not background, it should be set `reduce_zero_label=True`.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:284]: it is no proper

File: docs/en/advanced_guides/add_dataset.md [hunk 9]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
-test_dataloader = dict(dataset=dataset_A_test)
+## How to prepare your own dataset
 
+Preparation of your own dataset is divided into following steps:
+
+- Dataset preparation, generating correct data format and saving it to correct place like `./data/`.
+- Define your dataset in `./mmseg/datasets/` to register it in `DATASETS`.
+- Setting hyperparameters of training and validation in `./configs/_base_/datasets/`, such as data augmentation strategy, image crop size and image pixel normalization.
+
+Files needed to be modified are listed below:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:209]: Datasets should follow the below structure.

File: docs/en/advanced_guides/add_dataset.md [hunk 10]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+
+```none
+mmsegmentation
+   |
+   |- data
+   |     |- my_dataset                 # your dataset after data conversion
+   |- mmseg
+   |     |- datasets
+   |     |     |- __init__.py          # add your dataset class here
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:218]:
```suggestion
   |     |     |- __init__.py          # import your dataset class here
```

File: docs/en/advanced_guides/add_dataset.md [hunk 11]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+   |     |     |- my_dataset.py               ## define your dataset class
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:219]:
```suggestion
   |     |     |- my_dataset.py               ## implement your dataset class
```

File: docs/en/advanced_guides/add_dataset.md [hunk 12]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+   |     |     |- ...
+   |- configs
+   |     |- _base_
+   |     |     |- datasets
+   |     |     |     |- my_dataset_config.py      # config of your dataset
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:224]:
```suggestion
   |     |     |     |- my_dataset_config.py      # config file of your dataset
```

File: docs/en/advanced_guides/add_dataset.md [hunk 13]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+        palette = [[255, 255, 255], [0, 0, 255], [0, 255, 255], [0, 255, 0],
+           [255, 255, 0], ...]
+    )
+
+    # The formats of image and segmentation map are both .png in this case
+    def __init__(self, **kwargs):
+        super(MyDataset, self).__init__(
+            img_suffix='.png',
+            seg_map_suffix='.png',
+            reduce_zero_label=False, # reduce_zero_label is False because label 0 is background (first one in CLASSES above)
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:277]: the explanation of `reduce zero label` is no proper

File: docs/en/advanced_guides/add_dataset.md [hunk 14]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+
+`classes` and `palette` in `METAINFO` are defined in `./mmseg/dataset/my_dataset.py`, which are class names and BGR values of annotations, respectively.
+Besides,`palette` would only be used in prediction visulization, which would not affect process of training and validation.
+
+Specifically, if label 0 in segmentation map is not background, it should be set `reduce_zero_label=True`.
+
+After creating `./mmseg/dataset/my_dataset.py`后，it is also necessary to add it in `./mmseg/dataset/__init__.py`:
 
 ```python
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:289]:
```suggestion
```

File: docs/en/advanced_guides/add_dataset.md [hunk 15]
```diff
@@ -58,99 +107,266 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+    'MyDataset'
+]
+```
+
+### Set config file of dataset
+
+After defining your dataset, it is also necessary to define configs of your dataset `my_dataset_config.py` in `./configs/_base_/datasets/`, which would be used concurrently with other config parameters in training and inference.
+
+```python
+# Your dataset type defined in ./mmseg/datasets/__init__.py
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:303]:
```suggestion
# Class name defined in ./mmseg/datasets/my_dataset.py
```

File: docs/en/advanced_guides/add_dataset.md [hunk 16]
```diff
@@ -1,5 +1,46 @@
 # Add New Datasets
 
+## Data configuration
+
+`train_dataloader`, `val_dataloader` and `test_dataloader` in config file are the variables for data configuration, to define the arguments that are used in datasets and dataloaders.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:5]:
```suggestion
`train_dataloader`, `val_dataloader` and `test_dataloader` in config file are the variables for dataloader configuration, to define the arguments that are used in datasets and dataloaders.
```

File: docs/en/advanced_guides/add_dataset.md [hunk 17]
```diff
@@ -1,5 +1,46 @@
...
+        type=dataset_type,
+        data_root=data_root,
+        data_prefix=dict(
+            img_path='images/validation',
+            seg_map_path='annotations/validation'),
+        pipeline=test_pipeline))
+test_dataloader = val_dataloader
+```
+
+- The setting of `xxx_dataloader` works for model training, validation and testing that are by [`build and registry`](https://github.com/open-mmlab/mmengine/blob/master/docs/en/tutorials/registry.md) mechanism.
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:36]:
```suggestion
- `train_dataloader`, `test_dataloader`, and `val_dataloader` are input variables of `Runner` and work for model training, validation and testing that
```

File: docs/en/advanced_guides/add_dataset.md [hunk 18]
```diff
@@ -58,99 +99,264 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
 For example, suppose the original dataset is `Dataset_A`, to repeat it, the config looks like the following
 
 ```python
-dataset_A_train = dict(
+dataset_A_train_dataloader = dict(
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:102]: there is no necessary to config a `dataset_A_train_dataloader`, please remove it  and just keep dataset configuration

File: docs/en/advanced_guides/add_dataset.md [hunk 19]
```diff
@@ -58,99 +99,264 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+If the datasets you want to concatenate are in the same type with different annotation files, you can concatenate the dataset configs like the following.
 
-```python
-dataset_A_train = dict()
-dataset_B_train = dict()
-concatenate_dataset = dict(
-    type='ConcatDataset',
-    datasets=[dataset_A_train, dataset_B_train])
-```
+````
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:120]:
```suggestion

```

File: docs/en/advanced_guides/add_dataset.md [hunk 20]
```diff
@@ -58,99 +99,264 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+You may concatenate two `dataset`.
+
+  ```python
+    dataset_train_1 = dict(
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:124]: The intent is not required.

File: docs/en/advanced_guides/add_dataset.md [hunk 21]
```diff
@@ -58,99 +99,264 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+        dataset=dict(type='ConcatDataset', datasets=[dataset_train_1, dataset_train_2]))
+  ```
+
+  In this case, `ann_dir_1` and `ann_dir_2` are corresponding to `ann_file_1.txt` and `ann_file_2.txt`.
+````
+
+### Multi-image Mix Dataset
+
+We use `MultiImageMixDataset` as a wrapper to mix images from multiple datasets.
+`MultiImageMixDataset` can be used by multiple images mixed data augmentation
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:150]:
```suggestion
`MultiImageMixDataset` can be used for images mixed data augmentation
```

File: docs/en/advanced_guides/add_dataset.md [hunk 22]
```diff
@@ -58,99 +99,264 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+like mosaic and mixup.
 
-A more complex example that repeats `Dataset_A` and `Dataset_B` by N and M times, respectively, and then concatenates the repeated datasets is as the following.
+An example of using `MultiImageMixDataset` with `Mosaic` data augmentation:
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:153]: There is an example...

File: docs/en/advanced_guides/add_dataset.md [hunk 23]
```diff
@@ -58,99 +99,264 @@ We use `RepeatDataset` as wrapper to repeat the dataset.
...
+            type=dataset_type,
+            reduce_zero_label=False,
+            data_prefix=dict(img_path=data_root + "images/train",
+                             seg_map_path=data_root + "annotations/train"),
+            pipeline=[
+                dict(type='LoadImageFromFile'),
+                dict(type='LoadAnnotations'),
+            ]
+        ),
+        pipeline=train_pipeline)
```
Comment (MeowZheng) [inline docs/en/advanced_guides/add_dataset.md:190]: what is the difference between `pipeline` in line 185 and `pipeline` in line 190
